pkg/ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,120 @@
+package ratelimiter
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/argoproj/argo-cd/v2/util/env"
+	"golang.org/x/time/rate"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type AppControllerRateLimiterConfig struct {
+	BucketSize      int
+	BucketQPS       int
+	FailureCoolDown time.Duration
+	BaseDelay       time.Duration
+	MaxDelay        time.Duration
+}
+
+func GetAppRateLimiterConfig() *AppControllerRateLimiterConfig {
+	return &AppControllerRateLimiterConfig{
+		// global queue rate limit config
+		int(env.ParseInt64FromEnv("WORKQUEUE_BUCKET_SIZE", 500, 1, math.MaxInt64)),
+		int(env.ParseInt64FromEnv("WORKQUEUE_BUCKET_QPS", 50, 1, math.MaxInt64)),
+		// individual item rate limit config
+		// when WORKQUEUE_FAILURE_COOLDOWN is 0 per item rate limiting default disabled(default)
+		time.Duration(env.ParseInt64FromEnv("WORKQUEUE_FAILURE_COOLDOWN_NS", 0, 0, (24 * time.Hour).Nanoseconds())),
+		time.Duration(env.ParseInt64FromEnv("WORKQUEUE_BASE_DELAY_NS", 1000, 1, (24 * time.Hour).Nanoseconds())),
+		time.Duration(env.ParseInt64FromEnv("WORKQUEUE_MAX_DELAY_NS", (120 * time.Second).Nanoseconds(), 1*time.Millisecond.Nanoseconds(), (24 * time.Hour).Nanoseconds())),
```

Comment:
Per offline discussion, no need to go as high as 120s between reconciles. We're trying to reduce reconciles from hundreds of times per minute, to dozens or less. 

I think a 2 or 3 second cap will add sufficient protection to the controller without degrading the end user experience.

Reply:
updated to 3 seconds